### ISSUE TITLE
Core support for a producer that accepts pushed records (was: withThreads after from)

### ISSUE THREAD

**Issue by zivziv77** [awslabs/pocket-etl#2]
Hi

I dont see any withThreads options after from is it on purpose?

Thanks

**Comment by bmaizels** [comment]
from() only takes a producer stage (which right now is limited to an extractor). There is no withThreads options when creating an extractor. If you wish to run multiple extractors in parallel you should instantiate each instance and pass them in as collection of extractors (as described in #3 ).

It just occurred to me writing this answer that you could theoretically pass the same instance of the same extractor multiple times in a collection to the producer stage and that might be a way to achieve what you could be asking for. I haven't written a functional test for this pattern though, so I don't know if it will work as advertised.

If that's what you are needing to do and you want to give it a try I'd love to know how that turns out.

**Comment by bmaizels** [comment]
By the way, the approach I suggested above would be problematic for any extractor that initialized its own state (which is most of them). For example an SqlExtractor will execute the query once it initializes upon job execution. If you referenced that same extractor multiple times you would likely get an exception complaining that you were attempting to open an extractor that is already open (initialized).

In a way, when I originally thought about this I just didn't anticipate much need for parallelized identical extractors. In the scenarios I could think of you would want them to be partitioned or sharded in some way (for instance an SqlExtractor that has multiple instances in parallel that each query a different range from the table). There would be very little benefit having multiple identical extractors in this scenario when all the hard-work (the query execution) can only be performed once anyway.

It will be interesting to learn if people do have use-cases for this.

**Comment by zivziv77** [comment]
Hi,

Thanks for your answer.
I thought about it when having a consumer from queue lets say rabbitmq and I want to have better performance and have several consumers from a queue fetching messags at the same time. If I could have several extractors threads that would have been nice...
Anyway that is really great what you wrote here. I plan to contribute some pull requests in the future if you need.
Have you ever thought of a validate step? If I want to have a step after extractor to validate if certain fields are exist. Should I use the filterTransformer? 
Thanks a lot
Ziv

**Comment by bmaizels** [comment]
That makes sense. One of the challenges here is that an extractor is a fundamentally different kind of thing (a producer) to a transformer and a loader (a consumer). The interface looks a bit like an Iterator, and the extractor is not responsible for its own orchestration it waits to be pulled, it doesn't push. Because of this, your options are a bit more limited, I think if I wanted to solve this problem I'd implement a special kind of extractor that took a Supplier of extractors and a parallelism count and then it would instantiate the number of desired extractors from that supplier and would proactively pull records from those extractors, buffer them and have them serve them up when its own next() function is called.

The alternative is to write an extractor with parallelism in mind, make it threadsafe so concurrent calls to next() don't mess it up, and it can be opened and closed multiple times then reference that same instantiation as multiple extractors using the existing interface. I think I prefer the first option because it can be more generally applied to any existing extractor.

**Comment by bmaizels** [comment]
Wrt/ validation, you could extend the existing SqlExtractor or write a transformer to perform this function. Transformers can return emptyList to effectively filter a record out of the stream entirely.

**Comment by zivziv77** [comment]
I understand what you are saying.
Having a push option to an extractor would have been much better.... I can
have several extractors listen to rabbitmq pushing the messages to internal
java queue then run pocket-etl stream to consume these messages.
Is that more less what you mean?

**Comment by bmaizels** [comment]
Actually I remember now that what I described with the parallel buffered extractors is basically exactly what the stream does, which is why I decided we didn't need it. Sorry to go round in circles a bit it's been some time since I thought about all this stuff.

I think you can specifically solve the rabbitmq case without any of the complex solutions above by initializing multiple instances of the same type of extractor pointing to the same queue and simply pass them in as distinct multiple extractors to the existing PocketEtl interface.

**Comment by bmaizels** [comment]
I was weary about giving full control to an extract to exhaustively push records until finished in case it needed to be interrupted for example. The way it's written right now allows PocketEtl to retain control of the execution of the stream. That was a design decision for better or for worse. There is a scenario where the extractor needs to be interrupted, if a downstream stage throws an UnrecoverableStreamException.

**Comment by zivziv77** [comment]
But isn’t it a problem to use the push option to those  extractors? The
consumers will be blocked waiting for messages and not pulling those
messages.

Ziv

**Comment by bmaizels** [comment]
The stream producer orchestration thread runs independently of the consumers. The extractor will be continuously polled for records until the buffer of the downstream consumer is full and then it will block until the consumer has caught up. The consumers will pull records from the buffer until it's empty then they will block until the producer thread has produced some more records.

In the scenario where your consumers run faster than your producer the consumers will frequently block waiting for new records. In the scenario where your producer runs faster than your consumers, the producer will run free until the consumer buffer is filled and then it will frequently block waiting for the consumers to process records.

In both of these scenarios the extractors and transformers and loaders are completely shielded from what's going on here behind the scenes and will not directly observe this blocking behavior themselves. All of this is handled by the EtlProducer and EtlConsumer objects that are working behind the scenes to orchestrate the stream.

**Comment by bmaizels** [comment]
If you're interested in looking at exactly how this works, this is the relevant area:
[Producer code](https://github.com/awslabs/pocket-etl/tree/master/src/main/java/com/amazon/pocketEtl/core/producer)
[Consumer code](https://github.com/awslabs/pocket-etl/tree/master/src/main/java/com/amazon/pocketEtl/core/consumer)

In an actual stream there are usually multiply stacked consumer and producer layers all with different responsibilities. Again that detail is abstracted from the user of the library.

**Comment by zivziv77** [comment]
I think I didn’t explain myself correctly. I meant that all the examples of
extractors in pocket-etc are doing sort of pull meaning they iterate on
something. They don’t block waiting for someone to push the messages.
On rabbitmq it is better using the push option .
So can write an extractor that have such push option  in its extract
method? I see that in sqs for example you use pull option not push.

Ziv

**Comment by bmaizels** [comment]
Okay, now I think I see where you're coming from. To paraphrase:

'You'd like an extractor that acts as a call-back hook for some other process running somewhere that will make the call when it receives new data.'

If that's what you are asking for I can confirm that there is nothing in PocketEtl that directly supports that pattern today. This could be solved generically if you had an extractor that buffered pushed data and then served it up through the next() command. You'd have to make next() block for a period of time if there were no messages immediately available (buffered) before actually returning Optional.empty() because doing so would indicate to the stream that there was no work remaining to do and the stream would terminate.

Another way to think of it is to create an extractor that is built around a blocking queue and then independently of the stream have your listener publish records directly into that blocking queue.

**Comment by bmaizels** [comment]
We could also create an EtlProducer that fully implemented this pattern and create an interface around it in the EtlStream.  I think conceptually this is a new type of producer that is -not- an extractor.

**Comment by zivziv77** [comment]
Yes now I see you understand me...
I am a heavy user of logstash which is sort of an etl tool. Sometimes I
feel logstash is too much.
I want small compact etl tool like you did.
In logstash stream lives forever until you stop the process.

Last thing I can theoretically block the optional.empty forever until I get
messages ? Meaning theoretically never send optional empty?

**Comment by bmaizels** [comment]
Yes I believe you can block next() indefinitely, although you will prevent the stream from terminating if a previous message generates an UnrecoverableStreamException(). If you implemented this in the core of PocketEtl as a distinct EtlProducer implementation you could avoid this problem by orchestrating the calls to next() yourself and having them run on a timeout and then just loop indefinitely (or until a signal that interrupts the stream is detected).

**Comment by zivziv77** [comment]
I think the solution you have with a new producer is much better.
I will try to implement it.
Thanks a lot for your time!
If you need me to help in anything let me know.
Ziv

**Comment by bmaizels** [comment]
I'll leave the issue open then. Thanks a lot for taking it on. Good luck, give me a shout if you need help.

**Comment by zivziv77** [comment]
Hi,
I started to work on it. several notes I need your advise on:

- for the pushed record I called streamExtractor and the header is: void handleDelivery(Queue<T> queue)throws UnrecoverableStreamFailureException;

- I added an internal queue since theoretically the stream can be opened forever until the process is terminated. 
- each object will be inserted into the queue from the extractor.
- inside the StreamExtractorEtlProducer which is the new Extractor ETL producer , the produce looks like this:
public void produce() {
		  if (isClosed) {
	            throw new IllegalStateException("Attempt to run extractor after extractor was closed");
	        }

	        try (EtlProfilingScope ignored = new EtlProfilingScope(parentMetrics, "ExtractorProducer." + name + ".produce")) {          
	            streamExtractor.handleDelivery(internalQueue);	            
	            while(!isClosed) {	               
	                try {
	                	downstreamEtlConsumer.consume(EtlStreamObject.of(internalQueue.take()));	                  	                   
	                } catch (UnrecoverableStreamFailureException e) {
	                    throw e;
	                } catch (RuntimeException e) {
	                   // nonFatalExceptionWasThrown = true;
	                   
	                } catch (InterruptedException e) {
						
						e.printStackTrace();
					}
	            } 
	        }

basically until the close is hit it is trying to take objects from the queue in a blocked way.
Now few questions that I need your opinion on:

-  in the StreamExecutorEtlProducer in the produce method where  the thread pool is, in a stream way the thread is endless, I think there should be number of thread in the pool like the number of streamExtractor. since each thread should handle one streamExtractor forever.

- Do you think there should be a common interface for batchExtractor and streamExtractor?

- I think you should change the name of extractor to be batchExtractor....

- in EtlExtractStage the code is almost the same but the extractor interface is different either batch or stream. How do you think that should be implemented?

Thanks

**Comment by bmaizels** [comment]
Hey, lots to unpack there. Let me start from a place of 'if I were doing this what would I probably do' then try and reconcile that with your ideas and questions and see where we land.

What's interesting about this problem is that conceptually this is a very different model to the existing extractor. There's literally no equivalent to an extractor object here because its the underlying producer object that needs to consume the objects there's no place for an object or interface that serves these objects, instead the producer needs to provide an interface that allows other objects running outside of PocketETL to push objects to.

So I imagined it would be something like this:
```
Consumer<MyStreamObject> streamConsumer = EtlStream.createStream().then(load(myLoader)).run();
```
The interface has taken a completely different path here. run() now returns a Consumer<MyStreamObject> that should be linked to a new Producer implementation in PocketETL that attempts to store the message on a blocking queue that is being consumed by the downstream consumer.

Now, all that being said I'll try and answer your questions directly:

>>in the StreamExecutorEtlProducer in the produce method where the thread pool is, in a stream way the thread is endless, I think there should be number of thread in the pool like the number of streamExtractor. since each thread should handle one streamExtractor forever.

I think it's going to be a single blocking queue. Multiple threads might be trying to put things into it. The downstream consumer will be trying to read things off of the queue, and could have one or multiple threads (but that's up to the downstream consumer and not a concern of the 'StreamExecutorEtlProducer'.

>> Do you think there should be a common interface for batchExtractor and streamExtractor?

No I think these are entirely different. I don't think that there should be such a thing as BatchExtractor (read my ideas above).

>> I think you should change the name of extractor to be batchExtractor....

I think you're right that it more accurately describes what it is, however I don't think the stream thing will contend with it so there's no reason to change it (as that will require a major version bump).

>> in EtlExtractStage the code is almost the same but the extractor interface is different either batch or stream. How do you think that should be implemented?

As per my ideas above I think we want to end up with an interface that looks something like this : 
```
Consumer<MyStreamObject> streamConsumer = EtlStream.createStream().then(load(myLoader)).run();
```

If you like you can just write up the interface proposal (that's where I usually start when I'm coding for myself), and put 'throws new UnsupportedOperationException();' in all the methods. I think it's easier to discuss proposals when looking at code.

**Comment by bmaizels** [comment]
One thing that occured to me is the StreamEtlProducer isn't going to have any threads of its own. Things are put into its queue from threads running outside the EtlStream, and threads running on the downstream consumer will be pulling things off its queue. It has no need for any threads itself, it's just a conduit between the outside world and the first consumer on the stream, or am I missing something?

**Comment by zivziv77** [comment]
Thanks for your answer.
I thought of doing like sqsExtractor SNSStreamExtractor that wait for
messages being pushed. I thought it should be part of the application.
I will write something and let you review it.
Thanks

**Comment by zivziv77** [comment]
One more thing.
I think that the in you example I cannot combine both stream and batch
extractors no?

**Comment by bmaizels** [comment]
> I thought of doing like sqsExtractor SNSStreamExtractor that wait for messages being pushed

Yeah this is the idea we first talked about, as I understand it a slight variation on the ExtractorProducer that doesn't infinitely block. I'll be honest though, I prefer the design where we completely outsource the producer execution to the thing pushing the messages (RabbitMQ?) and just leave an opening at the head of the stream for it to push messages to the first consumer, in my mind it's simpler and involves less moving parts (the complexity is on the interface side) and I always opt for simplicity :) Do you really prefer to shim this through an extractor-type object?

SQS does not support pushing messages, so I don't think there is really a different way to do that particular one. For SNS, it would be tricky to get that to push messages to your stream, I think you'd have to do it through an open TCP socket (with HTTP delivery), in that instance you'd have a thread listening on the socket and would then push messages as it received them into the stream. I think it's going to be more preferable to most people to link their SNS topic to an SQS queue and poll it or link it to a lambda to react and do things (without a stream at all). We could adapt the SQSExtractor to be more stream-like and less batchy, although the batch functionality needs to remain because that supports existing use-cases.

I'll take a look at RabbitMQ at some point so I can understand better the problem you are specifically trying to solve.

**Comment by bmaizels** [comment]
>I think that the in you example I cannot combine both stream and batch extractors no?

That's a really good point actually and undermines my suggestion. Your idea wins on those grounds :) So.... We need some kind of extractor-like object that is not an extractor. It needs a consumer method on it that other threads can use to push objects into it. All it really needs to do is take those objects and push them onto the queue of the downstream consumer. It'd be great if we could think of a way to just have it directly wired to the downstream consumer rather than have a thread who's only job it is to move messages from one queue to another.

**Comment by zivziv77** [comment]
You just sent me this mail when I was thinking that you way is the best...
so you dont think we need an extractor something like that? what you is
saying that the extractor like will get an object from outside of
pocket-etl and will jus move them to the downstream?


void run(Queue<T> queue)throws UnrecoverableStreamFailureException;
default void open( EtlMetrics parentMetrics) {
}
@Override
    default void close() throws Exception {
        //no-op
    }

can you just suggest other interface....?

thansk
Ziv

**Comment by bmaizels** [comment]
Hmm, how about something like this (forgive small errors I'm writing code without an IDE):

```
public interface StreamConsumer<T> extends Autoclosable, Consumer<T> {
   void open(EtlMetrics parentMetrics);
}
```

So we will just use the standard Java consumer interface so it's clear what this thing is doing. Some documentation to explain that the call to accept() might block if the stream is backed up.

On the EtlStream side it will be something like :

```
// I think we can overload extract here
EtlStream.extract(myStreamConsumer).then(...).run();
```

I don't know whether to call it a StreamConsumer or a StreamProducer. Literally, and from the context of being outside the stream, it's a Consumer, but from the context of within the stream it's a producer. What do you think makes more sense to someone using the library?

**Comment by zivziv77** [comment]
Hi
I think it is more producer no?
How do you think the method accept should looks like?

**Comment by bmaizels** [comment]
I think maybe we should avoid using the term 'Producer' or 'Consumer' for this interface then as it will likely cause more confusion than be helpful. If I think if something good I'll let you know.

The accept() method comes from the standard Java 'Consumer' interface which my suggestion extended. I think we should leverage standard patterns like that where possible so people aren't surprised and it also maximizes compatibility and interoperability between libraries..

**Comment by zivziv77** [comment]
ok I agree consumer interface is the best suit here.
so basically accept will get an object and put it in the downstream
consumer thread pool queue?

I couldn't find that the downstream expose its queue beisde here:
public class EtlExecutorFactory {
    /**
     * This multi-threaded EtlExecutor uses a fixed-size work queue that
will block on new requests once the queue is
     * full until it drains enough to add the new task.
     * @param numberOfWorkers Number of threads to run tasks simultaneously.
     * @param queueSize The maximum size of the work-queue. Submit will
block once this hits its size limit.
     * @return A fully constructed EtlExecutor.
     */
    public EtlExecutor newBlockingFixedThreadsEtlExecutor(int
numberOfWorkers, int queueSize) {
        ExecutorService executorService = new
ThreadPoolExecutor(numberOfWorkers, numberOfWorkers, Long.MAX_VALUE,
                TimeUnit.NANOSECONDS, new ArrayBlockingQueue<>(queueSize),
(runnable, threadPoolExecutor) -> {
                    if (threadPoolExecutor.isShutdown()) {
                        throw new
RejectedExecutionException("ExecutorService was shutdown");
                    }
                    try {
                        threadPoolExecutor.getQueue().put(runnable);
                    } catch (InterruptedException ignored) {
                        throw new RejectedExecutionException("Thread was
interrupted trying to queue new work");
                    }
                });

        return new ExecutorServiceEtlExecutor(executorService);
    }

but still the internal queue is not exposed am I right?

**Comment by bmaizels** [comment]
Yes, what you found there is factory for creating an Executor that uses a fixed size blocking queue for its work input. This component is used by all the EtlConsumer classes that do actual work.

What we need on our StreamExtractor is some kind of blocking queue that can interrupt a blocked thread if an UnrecoverableStreamException is returned by any message sent to the downstream consumer. I don't have a solution in mind it will require a bit of deep thought. Let me know if you have any ideas how to accomplish this. If we can figure out a way to do this without actually running an executor that would be ideal, but maybe due to the interruptability constraint it's not possible.

**Comment by zivziv77** [comment]
at the beginning I thought of putting jobs directly in the executer service
queue.
Do you know if is possible?

Ziv

**Comment by bmaizels** [comment]
As long as we don't expose the executor directly, and as long as we can guarantee that the queue will not block indefinitely if the consumer throws or receives an UnrecoverableStreamException I think that'll work.

**Comment by zivziv77** [comment]
Sorry I was away for few days.
So just to summarize what we need to do :
1.either create a queue or use executed service queue.
2. Thread listen in block mode to that queue and forward event to executer.
3. We also need not to terminate stream and keep it runners by forever or nail we explicitly stop it.

What do you think?

Ziv

**Comment by bmaizels** [comment]
Hi, welcome back!

If we can find a way to not have to do number 2 that would be great. The downstream consumer already has a blocking queue and an executor working on it, if we could just have a way for the producer to directly put objects into that blocking queue then there's no need to have a dedicated thread to just move messages from one queue to another.
